### DISCUSSION TITLE
Why does using TTL not work?

### DISCUSSION THREAD

**Discussion by LiJie20190102** [JanusGraph/janusgraph#4346] [Q&A] [answered]
Here is my code. After a long period of time, I can still query the data. Is there a problem with my code?

` 

    public void testTTL() {
        JanusGraph graph = ConfiguredGraphFactory.open(GRAPH_NAME);
        JanusGraphManagement mgmt = graph.openManagement();

        PropertyKey sensor = mgmt.makePropertyKey("sensor").cardinality(Cardinality.SINGLE).dataType(Double.class).make();
        mgmt.setTTL(sensor, Duration.ofSeconds(50));

        PropertyKey sensor2 = mgmt.makePropertyKey("sensor2").cardinality(Cardinality.SINGLE).dataType(Double.class).make();
        mgmt.setTTL(sensor2, Duration.ofSeconds(100));

        // vertex ttl , only support static vertex
        VertexLabel static_tweet = mgmt.makeVertexLabel("static_tweet").setStatic().make();
        mgmt.addProperties(static_tweet, sensor, sensor2);
        mgmt.setTTL(static_tweet, Duration.ofSeconds(100));

        VertexLabel tweet = mgmt.makeVertexLabel("tweet").setStatic().make();
        mgmt.addProperties(tweet, sensor, sensor2);

        // edge ttl
        EdgeLabel visits = mgmt.makeEdgeLabel("visits").make();
        mgmt.addProperties(visits, sensor, sensor2);
        mgmt.addConnection(visits, static_tweet, tweet);
        mgmt.setTTL(visits, Duration.ofSeconds(100));

        // index
        mgmt.buildIndex("static_tweet", Vertex.class).addKey(sensor).addKey(sensor2).indexOnly(static_tweet).buildMixedIndex(GRAPH_NAME);
        mgmt.buildIndex("tweet", Vertex.class).addKey(sensor).addKey(sensor2).indexOnly(tweet).buildMixedIndex(GRAPH_NAME);
        mgmt.buildIndex("visits", Edge.class).addKey(sensor).addKey(sensor2).indexOnly(visits).buildMixedIndex(GRAPH_NAME);
        mgmt.commit();
    }

  
    public void testAddTTLData() {
        JanusGraph graph = ConfiguredGraphFactory.open(GRAPH_NAME);
        JanusGraphTransaction tx = graph.buildTransaction().checkInternalVertexExistence(false).start();
        // static不能设定顶点id
        JanusGraphVertex staticTweet = tx.addVertex(T.label, "static_tweet", "sensor", 1.5, "sensor2", 2.5);
        JanusGraphVertex tweet = tx.addVertex(T.label, "tweet", "sensor", 1.6, "sensor2", 2.7);

        JanusGraphEdge visits = staticTweet.addEdge("visits", tweet);
        visits.property("sensor", 3.3);
        visits.property("sensor2", 3.4);

        tx.commit();
    }

  
    public void getTTLData() {
        JanusGraph open = ConfiguredGraphFactory.open(GRAPH_NAME);
        JanusGraphTransaction tx1 = open.newTransaction();

        Iterator<Vertex> iterator = tx1.vertices();
        while (iterator.hasNext()) {
            Vertex next = iterator.next();
            VertexProperty<Object> sensor = next.property("sensor");
            String key = sensor.key();
            Object value = sensor.value();
            System.out.println(next.id() + "===1===" + key + "===" + value);
            VertexProperty<Object> sensor2 = next.property("sensor2");
            String key2 = sensor2.key();
            Object value2 = sensor2.value();
            System.out.println(next.id() + "===2===" + key2 + "===" + value2);

        }

        Iterator<Edge> edgeIterator = tx1.edges();
        //    Iterator<Edge> edgeIterator = tx1.edges();
        Map<String, Object> hashMap = new HashMap<>();
        if (edgeIterator.hasNext()) {
            System.out.println("====1");
            Edge next = edgeIterator.next();
            next.properties()
                .forEachRemaining(
                    x -> {
                        String key = x.key();
                        Object value = x.value();
                        hashMap.put(key, value);
                    });
        }
        System.out.println(hashMap);
    }
`

**Comment by li-boxuan** [comment]
What storage backend are you using?

**Reply by LiJie20190102** [reply]
Hbase. When I checked the edges just now, there was no data left (but it was far beyond the time I set), but there is still data for the vertices. Meanwhile, my edge connects two types of vertices, one is static with ttl set, and the other is not static

**Reply by li-boxuan** [reply]
100 seconds as TTL is too small and generally an anti-pattern in HBase/Cassandra. But yeah it should still work... what if you only add vertex without edge attached?

**Reply by LiJie20190102** [reply]
Now there are only vertex points and no edges in the graph, but the vertices will not expire

**Reply by LiJie20190102** [reply]
I previously set it to 5 minutes, but it still doesn't work

**Reply by li-boxuan** [reply, marked as answer]
I just ran testVertexTTLWithCompositeIndex test with a local HBase container. The test worked fine. Would you be able to rewrite your experiment in a single test format?

**Reply by LiJie20190102** [reply]
![image](https://github.com/JanusGraph/janusgraph/assets/53458004/6196134b-9616-4279-9eae-d24bd470aeac)
Still not deleted. Is it because we modified the code that caused the problem? My version is 1.0.0

**Reply by li-boxuan** [reply]
What code did you modify?

**Reply by LiJie20190102** [reply]
I will check and modify it myself. Thank you for your help
